julia/README.md

```diff
@@ -75,16 +75,22 @@ julia> typeof(juliaize(m.x))
 Int32
 ```
 
-It is not presently possible to `juliaize` arrays (but you can get them as the output of functions, see below).
+It is not presently possible to `juliaize`/`dexize` arrays (but you can use them / get them as the input/output of `NativeFunction`s, see below).
```

Comment:
We should implement it... What kind of array representation does Julia use? Are they strided arrays? Or contiguous Fortran-ordered arrays?

Reply:
Not particularly important, and not blocking for this PR.
But yes, at some point.
I am not sure what a strided array is in this context.
I would have counted a contiguous Fortran ordered array as being a type of strided array.
(and indeed in Julia the default `Array` type is a subtype of the `StridedArray` union, along with many kinds of views, and transposes)

But regardless, yes the default `Array` type is a contiguous Fortran ordered array.
Probably for initial support of this can do a contiguous `C` ordered array, and can take care of the reordering in the API.
Like we do when passing `Array`s to `NativeFunction`s.

Reply:
By strided array I mean that an array is represented as:
1. a data pointer
2. an integer describing rank
3. an array of integers (of length matching rank) describing its shape (extent)
4. an array of integers (of length matching rank) describing the stride in each dimension. That is, `strides[i]` is the increment that should be applied to a data pointer to traverse to the next element along dimension `i`.

Strided array representation is a generalization of C arrays, Fortran arrays, and it can handle various views and transposes without requiring data copies. It's what NumPy and PyTorch use for example and it's extremely widespread.

Since Julia seems to be using it too, I think it makes sense to support this as the general exchange method.

Reply:
I am not sure, let me summon someone who actually understands how arrays are implemented under the hood.
(Not really part of this PR, but since we have started talking here)

Reply:
@StefanKarpinski says:

> Adam’s definition of a strided array seems to be accurate, but Julia standard `Array`s are not that general: they have contiguous memory and strides are always in column-major order (last-index contiguous)
> We expose more complex layouts through other array types which ultimately all share memory with a contiguous `Array` object

My summary of that is: there is no type in julia that has a memory layout with that `"1. a data pointer 2. an integer describing rank ..."`
A plain `Array` in julia is a just a Fortran array.
and other layouts use other types, like the `PermutedDimsArray`.

Probably a pretty cute little package could be created that does match that layout and provides nice helpers to convert to and from various julia `AbstractArray` types using stdlib, and/or packages like @Jutho's Strided.jl [StridedView](https://github.com/Jutho/Strided.jl/blob/fe8db2cd9d0fae9b08db08eb4858b0f45ace6f23/src/stridedview.jl#L2-L8)
With the resolve being generally zero-copy.
I am not presently aware of one.

Reply:
I think my `StridedView` is exactly of that type described by @apaszke . It has a rank, a size or extent for each dimension, and a stride in each dimension. There is no particular restriction on the strides being ordered, or the smallest stride being 1, or any of that. Hence, we can compute with each entry of the array a corresponding "memory position" where that entry is to be found. The only difference is that, instead of directly getting data from memory, it assumes that all these "memory positions" correspond to the memory of an underlying (contiguous, as pointed out by Stefan) julia `Array`.

Reply:
Does it have the right layout that we can directly use it in a `ccall` ?

Reply:
Unlikely, but I have to admit that I just commented on this specific question, without actually knowing what this discussion or even this repository is about. 
The definition is here:
https://github.com/Jutho/Strided.jl/blob/master/src/stridedview.jl#L2

It furthermore has an `offset` (so that the first entry of the `StridedView` is not necessarily the first entry of the underlying parent array), and an (involutive) operation like `conj` that can can be applied on the fly.

Reply:
Ok, so since strided arrays are a generalization of contiguous arrays, I'm thinking of just making those the canonical representation. You can always derive the strides of a fortran array from its sizes, so you should be able to do ccalls with both native Julia arrays and the strided extensions.

Reply:
Well it depends. We have a choice between specializing on the particular input and compiling for that, or emitting code that will be generic over all possible configurations of sizes and strides. But flexibility in that is a good thing, we can always try to figure out what makes more sense based on how people end up using this.

Also, I'm pretty sure that the indexing expression you mentioned actually is representable with strided arrays, but you do need the offset pointer that I skipped and @Jutho mentioned. But of course there are always edge cases, like indexing with another array, which will not be representable as strided arrays.